index.js

```diff
@@ -50,51 +52,40 @@ module.exports = function (filepath, options, callback) {
         childArgs.push(JSON.stringify(opt.page));
     }
 
-    proc = childProcess.execFile(binPath, childArgs, function (err, stdout, stderr) {
-        var out,
-            result,
-            message,
-            output;
-
-        console.log('Testing ' + chalk.blue(path.relative(__dirname, filepath)));
+    console.log('Testing ' + chalk.blue(testUri));
 
-        if (stdout) {
-            try {
-                stdout.trim().split('\n').forEach(function (line) {
-                    var test;
+    // phantomjs [phantomjs-options] runner testuri [timeout [page]]
+    proc = childProcess.spawn(binPath, childArgs);
 
-                    try {
-                        out = JSON.parse(line.trim());
-                        result = out.result;
-
-                        message = 'Took ' + result.runtime + ' ms to run ' + result.total + ' tests. ' + result.passed + ' passed, ' + result.failed + ' failed.';
-
-                        output = result.failed > 0 ? chalk.red(message) : chalk.green(message);
+    proc.stdout.on('data', function (data) {
+        var out,
+            test,
+            message,
+            line = data.toString().trim();
 
-                        console.log(output);
+        try {
+            out = JSON.parse(line);
+        } catch (err) {
+            console.log(line);
+            return;
+        }
 
-                        if (out.exceptions) {
-                            for (test in out.exceptions) {
-                                console.log('\n' + chalk.red('Test failed') + ': ' + chalk.red(test) + ': \n' + out.exceptions[test].join('\n  '));
-                            }
-                        }
-                    } catch (e) {
-                        console.log(line.trim());
-                    }
-                });
-            } catch (e) {
-                this.emit('error', new Error(e));
+        if (out.exceptions) {
+            for (test in out.exceptions) {
+                console.log('\n' + chalk.red('Test failed') + ': ' + chalk.red(test) + ': \n' + out.exceptions[test].join('\n  '));
             }
         }
 
-        if (stderr) {
-            console.log(stderr);
-        }
+        if (out.result) {
+            message = 'Took ' + out.result.runtime + ' ms to run ' + out.result.total + ' tests. ' + out.result.passed + ' passed, ' + out.result.failed + ' failed.';
 
-        if (err) {
```

Comment:
@haiiaaa question, it appears there is no equivalent in `childProcess.spawn` to this error in the callback of `childProcess.execFile`. Test failures will cause that error to be thrown and then you can show different messages if there are test failures versus no test failures. Is there another way to achieve the same thing with spawn?

Reply:
According to the documentation this should be possible:

    proc.on('error', (err) => {
      console.log(err);
    });

Reply:
Yeah, and in es5 that would be 

    proc.on('error', function (err) {
      console.log(err);
    });

Would you like a PR?

Reply:
Yeah, I tried that and it does not do what I am looking for. It only errors in that case if the process fails to spawn somehow.

Reply:
Ok, now I see that here: https://nodejs.org/api/child_process.html#child_process_event_error

By test failures, do you mean failures in `out.exceptions` or do errors thrown from the javascript code?

Reply:
I think the qunit runner throws a javascript error if the test fails.

Reply:
I guess you want the following as output to the console if any test fails, right? I have not been able to achieve it with spawn.

    { Error: Command failed: C:\ProgramData\chocolatey\bin\phantomjs.EXE C:\dev\github\test\node_modules\qunit-phantomjs-runner\runner-json.js ./tests.htm 120
    
        at ChildProcess.exithandler (child_process.js:270:12)
        at emitTwo (events.js:130:20)
        at ChildProcess.emit (events.js:213:7)
        at maybeClose (internal/child_process.js:927:16)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
      killed: false,
      code: 1,
      signal: null,
      cmd: 'C:\\ProgramData\\chocolatey\\bin\\phantomjs.EXE C:\\dev\\github\\test\\node_modules\\qunit-phantomjs-runner\\runner-json.js ./tests.htm 120' }

If this is a showstopper it is also possible to stream with execFile but then the buffer would have to be specified manually for large buffers.

     buffer = options.buffer || 200; // default value
     proc = childProcess.execFile(binPath, childArgs, { maxBuffer: buffer*1024 }, function (err){
        if (err) {
            console.log(err);
        }
    });

Reply:
I found a solution. Just wondering if you knew of anything. Thanks.